### PR TITLE
Implemented restricted random fourier features

### PR DESCRIPTION
**Pull request by HedwigNordlinder**
Implemented restriction above the main diagonal. This means all values above the main diagonal are multiplied by an interpolation parameter, sampled normally before pushed through a function mapping it to [0,1]

### CODE REVIEW

File: src/smoothFUBAR/restrictedFUBAR.jl [hunk 1]
```diff
@@ -0,0 +1,142 @@
+struct LogPosteriorRFF{T} # Quite sure this is not needed since we already have it in smoothFUBAR.jl
```
Comment (murrellb) [inline src/smoothFUBAR/restrictedFUBAR.jl:1]: Drop this, as you say.

File: src/smoothFUBAR/restrictedFUBAR.jl [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    matrix_dimension = Integer(sqrt(length(flat_unrestricted_thetas)))
+
+    unrestricted_thetas = reshape(flat_unrestricted_thetas,matrix_dimension, matrix_dimension)
+    
+    interpolated_ψ = softmask(ψ)
+
+    lower_restriction_matrix = (UpperTriangular(reshape(ones(matrix_dimension^2),matrix_dimension,matrix_dimension))) 
+    upper_restriction_matrix = LowerTriangular(reshape(interpolated_ψ .* ones(matrix_dimension^2),matrix_dimension,matrix_dimension)) .- diagm(interpolated_ψ .* ones(matrix_dimension)) # So that we do not OW the diag
+    restriction_matrix = lower_restriction_matrix + upper_restriction_matrix
+
```
Comment (murrellb) [inline src/smoothFUBAR/restrictedFUBAR.jl:30]: Use LP.posmask for all of this. It is already flat, matching the theta vec.

File: src/smoothFUBAR/restrictedFUBAR.jl [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    return softmax(restricted_matrix[:])
+
+end
+
+function restricted_LL(ω,ψ, LP::LogPosteriorRFF)
+    theta_vec = restricted_thetas(ω,ψ, LP)
+    return sum(log.(theta_vec' * LP.condlik))
+end
+
+#=
```
Comment (murrellb) [inline src/smoothFUBAR/restrictedFUBAR.jl:42]: Remove these legacy commented bits.

File: src/smoothFUBAR/restrictedFUBAR.jl [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+    gridcoords = Matrix{Float64}(hcat(alpha_ind_vec, beta_ind_vec)')
+    grid_values = beta_vec[1:Int(sqrt(length(beta_vec)))]
+    
+    ωeights = randn(K)
+    initial_ψ = randn()
+    initial_θ = (ω = ωeights, ψ = initial_ψ)
+    flat_initial_θ, unflatten = value_flatten(initial_θ)
+    println(length(flat_initial_θ))
+    # K + 1 is almost certainly wrong
```
Comment (murrellb) [inline src/smoothFUBAR/restrictedFUBAR.jl:92]: You are incorrect about being incorrect!

File: src/smoothFUBAR/restrictedFUBAR.jl [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    verbosity > 0 && println("Step 4: Estimating posterior surface by HMC.")
+
+    gridcoords = Matrix{Float64}(hcat(alpha_ind_vec, beta_ind_vec)')
+    grid_values = beta_vec[1:Int(sqrt(length(beta_vec)))]
+    
+    ωeights = randn(K)
+    initial_ψ = randn()
+    initial_θ = (ω = ωeights, ψ = initial_ψ)
+    flat_initial_θ, unflatten = value_flatten(initial_θ)
+    println(length(flat_initial_θ))
```
Comment (murrellb) [inline src/smoothFUBAR/restrictedFUBAR.jl:91]: Remove print.

**Review by murrellb — CHANGES_REQUESTED** [review]

**Comment by murrellb** [comment]
In addition to the comments above, we need a better name for the user-facing function.

**Comment by murrellb** [comment]
Also, you need to add:
- a trace plot for ψ
- a printed (and logged, to a .txt file) statement of p(ψ>0) counted off the MCMC chain (which is our alignment-wide test)

File: src/smoothFUBAR/restrictedFUBAR.jl (hunk 1 above)
Comment (HedwigNordlinder) [reply src/smoothFUBAR/restrictedFUBAR.jl]: We now allow the user to choose a prior for psi other than a normal one, so new struct was created to

File: src/smoothFUBAR/restrictedFUBAR.jl (hunk 2 above)
Comment (HedwigNordlinder) [reply src/smoothFUBAR/restrictedFUBAR.jl]: Done

File: src/smoothFUBAR/restrictedFUBAR.jl (hunk 3 above)
Comment (HedwigNordlinder) [reply src/smoothFUBAR/restrictedFUBAR.jl]: Done

File: src/smoothFUBAR/restrictedFUBAR.jl (hunk 5 above)
Comment (HedwigNordlinder) [reply src/smoothFUBAR/restrictedFUBAR.jl]: Done

**Comment by HedwigNordlinder** [comment]
Closing this and making a new one

**Pull request closed without merging**